hkdf/src/sealed.rs

```diff
@@ -1,84 +1,61 @@
 use hmac::digest::{
-    core_api::{BlockSizeUser, CoreWrapper, OutputSizeUser},
-    Digest, FixedOutput, KeyInit, Output, Update,
+    block_buffer::{BufferKind, Eager, Lazy},
+    core_api::{BlockSizeUser, BufferKindUser, CoreWrapper, OutputSizeUser},
+    Digest, FixedOutput, KeyInit, Update,
 };
 use hmac::{EagerHash, Hmac, HmacCore, SimpleHmac};
 
-pub trait Sealed<H: OutputSizeUser> {
-    type Core: Clone;
+static EXPECT_MSG: &str = "HMAC can take a key of any size";
 
-    fn new_from_slice(key: &[u8]) -> Self;
+pub trait Sealed: OutputSizeUser + Sized {
+    type FullHmac: KeyInit + Update + FixedOutput;
+    type CoreHmac: KeyInit;
 
-    fn new_core(key: &[u8]) -> Self::Core;
+    fn core_to_full(core: &Self::CoreHmac) -> Self::FullHmac;
 
-    fn from_core(core: &Self::Core) -> Self;
-
-    fn update(&mut self, data: &[u8]);
+    fn new_core(key: &[u8]) -> Self::CoreHmac {
+        Self::CoreHmac::new_from_slice(key).expect(EXPECT_MSG)
+    }
 
-    fn finalize(self) -> Output<H>;
+    fn new_full(key: &[u8]) -> Self::FullHmac {
+        Self::FullHmac::new_from_slice(key).expect(EXPECT_MSG)
+    }
 }
 
-impl<H> Sealed<H> for Hmac<H>
+impl<C, K> Sealed for CoreWrapper<C>
 where
-    H: EagerHash + OutputSizeUser,
+    K: HmacKind<Self> + BufferKind,
+    C: BufferKindUser<BufferKind = K> + OutputSizeUser,
```

Comment:
Seems like this would preclude switching to newtypes of `CoreWrapper` for the hash crates.

Could this blanket impl be based around the requisite traits instead?

Reply:
I don't think it's possible to do automatically without specialization. Alternatively, we could introduce yet another trait which would regulate switching between those two HMAC implementations, i.e. we would move this burden to all hash implementation crates.

Reply:
Then I would personally rather have newtypes than this automatic switching.

Have you benchmarked how much `SimpleHmac` actually degrades performance for the short inputs in HKDF usage?

Reply:
We already have the `SimpleHkdf` type aliases, which do the job just fine.

It's more about structure size, depending on hash, `SimpleHmac` can be twice bigger, especially if the `reset` feature is enabled for `hmac`.